### DISCUSSION TITLE
PostgreSQL with SSL support?

### DISCUSSION THREAD

**Discussion by OckhamOdyssey** [dani-garcia/vaultwarden#3291] [Q&A] [answered]
I'm starting a Vaultwarden Docker container, it connects with a PostgreSQL external instance, allowing only SSL connections.

I add this environment variable inside the container `DATABASE_URL=postgresql://user:password@0.0.0.0/vaultwarden?sslmode=require`.

When I start the container, I have this error:
```
[2023-03-02 11:28:00.819][vaultwarden::util][WARN] Can't connect to database, retrying: DieselCon.
[CAUSE] BadConnection(
    "FATAL:  connection requires a valid client certificate\n",
)
```
I have the `client.crt`, `client.key`, and `server-ca.pem`, but I'm not sure where I should add them so that Vaultwarden can use them. I read [this documentation](https://github.com/dani-garcia/vaultwarden/wiki/Using-the-PostgreSQL-Backend), but it said nothing about SSL connections; just normal connections, migrations, and MySQL SSL connections. It is supported?

**Comment by jjlin** [comment, marked as answer]
IIRC, the database URL ends up getting parsed by `libpq`:

https://www.postgresql.org/docs/current/libpq-envars.html

So you would need to specify the corresponding connection parameters: `sslcert`, `sslkey`, and `sslrootcert`. Remember that the paths need to be from the perspective of the Docker container, not the Docker host.

Specifying those values via the corresponding environment variables should work too, but I haven't tried that myself.

**Reply by OckhamOdyssey** [reply]
As I have set up this environment for testing purposes, I have taken the liberty of testing both options.

`postgresql://user:passsword@0.0.0.0/vaultwarden?sslmode=require&sslcert=/path/to/client-cert.pem&sslkey=/path/to/client-key.pem&sslrootcert=/path/to/server-ca.pem` works fine.

Inside the docker compose, I change/add these environments:
```
- DATABASE_URL=postgresql://user:password@0.0.0.0/vaultwarden?sslmode=require
- PGSSLCERT=/path/to/client-cert.pem
- PGSSLKEY=/path/to/client-key.pem
- PGSSLROOTCERT=/path/to/server-ca.pem
```
It also works fine. So thank you. Both ways solve this :)
